src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/TypePreinit.cs

```diff
@@ -1864,12 +1944,49 @@ private static BaseValueTypeValue NewUninitializedLocationValue(TypeDesc locatio
             }
         }
 
-        private bool TryHandleIntrinsicCall(MethodDesc method, Value[] parameters, out Value retVal)
+        private bool TryHandleIntrinsicCall(TypeSystemContext context, MethodDesc method, Value[] parameters, out Value retVal)
         {
             retVal = default;
 
             switch (method.Name)
             {
+                case "Memmove":
+                    if (method.OwningType is MetadataType spanHelpersType
+                        && spanHelpersType.Name == "SpanHelpers" && spanHelpersType.Namespace == "System"
+                        && spanHelpersType.Module == spanHelpersType.Context.SystemModule
+                        && parameters[0] is ByRefValue dest
+                        && parameters[1] is ByRefValue src
+                        && parameters[2] is ValueTypeValue len)
+                    {
+                        int length = context.Target.PointerSize == 8 ? checked ((int)len.AsInt64()) : len.AsInt32();
+                        var srcSpan = new Span<byte>(src.PointedToBytes, src.PointedToOffset, length);
+                        var dstSpan = new Span<byte>(dest.PointedToBytes, dest.PointedToOffset, length);
+                        srcSpan.CopyTo(dstSpan);
+                        return true;
+                    }
+                    return false;
+                case "IsReferenceOrContainsReferences":
+                    if (method.OwningType is MetadataType rtType
+                        && rtType.Name == "RuntimeHelpers" && rtType.Namespace == "System.Runtime.CompilerServices"
+                        && rtType.Module == rtType.Context.SystemModule
+                        && method.Instantiation.Length == 1)
+                    {
+                        var type = method.Instantiation[0];
+                        bool result = type.IsGCPointer || (type is DefType { ContainsGCPointers: true });
```

Comment:
in the coreclr's interpreter we do this:
```
    bool containsGcPtrs = typeArg->ContainsGCPointers();

    // Return true for byref-like structs with ref fields (they might not have them)
    if (!containsGcPtrs && typeArg->IsByRefLike())
    {
        containsGcPtrs |= HasByrefFields(typeArg);
    }
```
so have to be careful about byref-like structs

Reply:
I stole that code from here https://github.com/dotnet/runtime/commit/1e0c24330c4074de89d13bca21b3b670f051f6eb#diff-daf353351989087d61e8a1cb5e9c1434b14f16a5029cd38c8ce0037ed4685fa7L48 was that code wrong then?

Reply:
I guess so? it's only recently it became possible with `allow ref struct` feature